clients/venice-thin-client/src/main/java/com/linkedin/venice/client/schema/RouterBackedSchemaReader.java

```diff
@@ -300,6 +290,294 @@ private SchemaEntry fetchSingleSchema(String requestPath, boolean isValueSchema)
+        if (schemaId == supersetSchemaId) {
+          supersetSchemaIsPreferredSchema = true;
+        }
+        if (maxPreferredSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID || maxPreferredSchemaId < schemaId) {
+          maxPreferredSchemaId = schemaId;
+        }
+      }
+    }
+    // Based on preferred schema ID and superset schema ID, compute the latest value schema ID.
+    boolean hasSupersetSchema = supersetSchemaId != SchemaData.INVALID_VALUE_SCHEMA_ID;
+    boolean hasPreferredSchema = maxPreferredSchemaId != SchemaData.INVALID_VALUE_SCHEMA_ID;
+    final int latestSchemaId;
+    if (hasPreferredSchema) {
+      if (supersetSchemaIsPreferredSchema) {
+        latestSchemaId = supersetSchemaId;
+      } else {
+        latestSchemaId = maxPreferredSchemaId;
+      }
+    } else {
+      if (hasSupersetSchema) {
+        latestSchemaId = supersetSchemaId;
+      } else {
+        latestSchemaId = maxSchemaId;
+      }
+    }
+    if (latestSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID) {
+      return;
+    }
+    // Perform latest value schema entry update.
+    if (latestValueSchemaEntry.get() == null || latestSchemaId == supersetSchemaId
+        || latestSchemaId > latestValueSchemaEntry.get().getId()) {
+      latestValueSchemaEntry.set(valueSchemaEntryMap.get(latestSchemaId));
+    }
+  }
+
+  private SchemaEntry maybeUpdateLatestValueSchemaEntry() {
+    SchemaEntry latest = latestValueSchemaEntry.get();
+    if (latest == null || shouldRefreshLatestValueSchemaEntry.get()) {
+      /**
+       * Every time it sees latestValueSchemaEntry is null or the flag to update latest schema entry is set to true,
+       * it will try to update it once.
+       * The update is expensive, but we expect it to be filled after the first fetch, as each store should have at least
+       * one active value schema.
+       */
+      updateAllValueSchemaEntriesAndLatestValueSchemaEntry(false);
+      shouldRefreshLatestValueSchemaEntry.compareAndSet(true, false);
+      latestValueSchemaEntry.set(latestValueSchemaEntry.get());
+      latest = latestValueSchemaEntry.get();
+    }
+    return latest;
+  }
+
+  /**
+   * This method will fetch the value schema by ID, and potentially will update the cached schema map.
+   * If the schema:
+   * (1) Exists in the cached schema map, we will directly return it.
+   * (2) Does not exist in the cached schema map, we will perform one time single schema fetch to request it from router,
+   * and update the result in the cached schema map.
+   * (3) If the schema was queried previously and not found, we will:
+   * -- (a) Return not found, if forceRefresh flag == false.
+   * -- (b) Perform one time router query, update and return the result if forceRefresh flag == true.
+   */
+  private SchemaEntry maybeUpdateAndFetchValueSchemaEntryById(int valueSchemaId, boolean forceRefresh) {
+    if (forceRefresh) {
+      SchemaEntry entry = fetchValueSchemaEntryFromRouter(valueSchemaId);
```

Comment:
When should we force refresh when the schema id is available locally?

Reply:
Do you mean if the schema is available in local and it is not INVALID_SCHEMA, then we don't have to call it again?

Reply:
Yeah, I think given the fact that schema is static, and will not change its content, I think it makes sense.